### DISCUSSION TITLE
How to achieve robust integration test setup with ReqnRoll?

### DISCUSSION THREAD

**Discussion by vyruz1986** [reqnroll/Reqnroll#229] [Q&A]
So after hearing some high-level stuff about SpecFlow for the past years, and meaning to get to know it, yesterday I decided to look into it, and quickly arrived here after learning SpecFlow is dead (all hail ReqnRoll :))

I feel like this should be a feature request, but since I'm very new to this framework I wanted to gather some input first.

What I'm trying to do is writing integration tests, mayb ReqnRoll is not the correct solution for this, if so, please elaborate.
For these integration tests I need to setup quite some dependencies (e.g. RabbitMQ, MSSQL server, Redis cache, ...)
I usually write integration tests using xUnit's `IAsyncLifetime` interface which I implement somewhere appropriate, this allows me to implment an `ValueTask InitializeAsync()` method which will be run at the start of the unit tests which I can then use in combination with e.g. https://dotnet.testcontainers.org/ to spin up all those 3rd party dependencies and retrieve connections strings for those instances. 
I then usually construct classes that need those dependencies (e.g. EF Core's `DbContext` implementation of the application I'm testing) and store them as a property in my test class so I can then use them in my test cases.

Now I'm wondering how to achieve something similar with ReqnRoll, I'm aware about Hooks (which somewhat resemble xUnit's `IAsyncLifetime`). And I'm also aware of the ReqnRoll's DI system (or even 3rd party integrations like Microsoft's DI container). 

So exploring those options, I come to the following conclusions:
- Using the built-in context injection doesn't seem possible, as this only seems to support POCO's which don't have dependencies of their own (like .e.g EFCores `DbContext` requires `DbContextOptions<T>` to be passed from somewhere.
- Using Microsoft's DI container plugin doesn't seem possible, as it requires implementing a static, non-synchronous method marked by the `[ScenarioDependencies]` attribute. Since https://dotnet.testcontainers.org/ only supports `async` methods to start/stop their containers, this won't work
- The only option left seems to be to define all my dependency setup in step definition files (or putting it in [a Driver](https://docs.reqnroll.net/latest/guides/driver-pattern.html)) and using Hooks to spin up the dependencies and construct `DbContext` etc instances.

I can hardly believe I'm the first trying to write integration tests, or perhaps I'm going about this completely the wrong way, so I would be greatful to hear how other people have solved this.

**Comment by AdaskoTheBeAsT** [comment]
hi, are you using xUnit? i can write few words later today about some tricks how I deal with that situation

**Reply by vyruz1986** [reply]
I mostly use xUnit yes, but am used to NUnit and MSTest as well. I might also create a real-world example with an open source project I work on to demonstrate what I'm trying to achieve

**Reply by AdaskoTheBeAsT** [reply]
ok finally night ;) 

there are multiple options to do this what you wrote about - my solution may be not optimal and will be highly dependant on the needs
lets try to split it to pieces

1) globally initialization for whole test set - integration test for db for example

- in one of the project when I have to use EF I do like this:
    1) Hooks folder in which I have GlobalHooks.cs class and async methods marked with attribute [BeforeTestRun] and [AfterTestRun] - those methods operate on some static class in which I have Testcontainers.MsSql initialization
    
```cs
public static class DbContextInitializer
{
	private static readonly MsSqlContainer MsSqlContainer
		= new MsSqlBuilder().WithExposedPort(55111).WithPassword("somePass123!").Build();

	public static MyDbContext? Instance { get; private set; }

	public static async Task InitializeAsync()
	{
		await MsSqlContainer.StartAsync();
		var connectionString = MsSqlContainer.GetConnectionString();
		await using var connection = new SqlConnection(connectionString);
		CreateDb(connection);
		InitData(connection);
		var sqlConnectionStringBuilder = new SqlConnectionStringBuilder(connectionString)
		{
			InitialCatalog = "MyDb"
		};
		var optionsBuilder = new DbContextOptionsBuilder<MyDbContext>();
		optionsBuilder.UseSqlServer(sqlConnectionStringBuilder.ConnectionString);
		var currentUserInfo = GetTestUserInfo();
		Instance = new MyDbContext(optionsBuilder.Options, currentUserInfo);
	}

	public static Task DisposeAsync()
	{
		Instance?.Dispose();
		return MsSqlContainer.DisposeAsync().AsTask();
	}

	private static void CreateDb(SqlConnection connection)
	{
		var initScriptPath = Path.GetFullPath(Path.Combine("Scripts", "init.sql"));
		var server = new Server(new ServerConnection(connection));
		var content = ScriptProcessor.PreprocessInitScript(initScriptPath);
		server.ConnectionContext.ExecuteNonQuery(content);
	}

	private static void InitData(SqlConnection connection)
	{
		for (int i = 1; i <= 20; i++)
		{
			var initDataScriptPath =
				Path.GetFullPath(Path.Combine($"Scripts", $"data_{i.ToString().PadLeft(2, '0')}.sql"));
			var server = new Server(new ServerConnection(connection));
			var content = ScriptProcessor.PreprocessInitDataScript(initDataScriptPath);
			server.ConnectionContext.ExecuteNonQuery(content);
		}
	}
}
```

- in above sample I have db schema initialization and data initialization script - in my case we have regular sql server project from which init script is generated (by sql server project publish) and then it is used to initialize whole db (we have views and other db things like table value parameters etc so safer is not to use db mmigration - there is preprocessor to trim a little init script
- and then dbContext is exposed as static property

2) in my other open source project I can't use global hook as I want to have few collections of tests in same project which will use different databases

in such case I use xunit collection fixtures (https://github.com/AdaskoTheBeAsT/AdaskoTheBeAsT.Identity.Dapper/tree/feature/insert-own-id/test/integ/AdaskoTheBeAsT.Identity.Dapper.SqlServer.IntegrationTest/TestCollections) unfortunatelly Reqnroll does not properly operates on IAsyncLifetime so I have to use regular constructor and Dispose with .GetAwaiter().GetResult() on async initialization and dispose

**Comment by ajeckmans** [comment]
If you want you can take a look at https://github.com/ajeckmans/ReqnrollIntegrationTests

This more or less is a way more simplified version of how I typically set up tests for a web application. I do want to note that it involves some less than typical set up of reqnroll and that I have not really organized the files nicely. I just did the bare minimum to create a webapp that has a dbcontext and a test set up that spins up a test container.

The reqnroll set up requires no packages other than the core reqnroll xunit package. I've found that the other DI containers get in the way and that BoDi is actually good enough for even advanced scenarios.

**Reply by vyruz1986** [reply]
I'll take a look later, thanks!

**Comment by ajeckmans** [comment]
@vyruz1986 Did any of our answers help?

**Comment by gasparnagy** [comment]
@vyruz1986 I would be interested to hear more about the comment you have made earlier

> Using Microsoft's DI container plugin doesn't seem possible, as it requires implementing a static, non-synchronous method marked by the [ScenarioDependencies] attribute. Since https://dotnet.testcontainers.org/ only supports async methods to start/stop their containers, this won't work

My understanding was that the method with `[ScenarioDependencies]` is only configuring the container, it does not need to "start" it... Could you please give some examples (even just pseudo code) that shows what is missing here?
